File: frontend/src/api/buildingComparison.ts
import { apiClient } from '@/api/client';

export interface BuildingComparisonEntry {
  building_id: string;
  building_name: string;
  address: string;
  passport_grade: string | null;
  passport_score: number | null;
  trust_score: number | null;
  completeness_score: number | null;
  readiness_summary: Record<string, boolean>;
  open_actions_count: number;
  open_unknowns_count: number;
  contradictions_count: number;
  diagnostic_count: number;
  last_diagnostic_date: string | null;
}

export interface BuildingComparison {
  buildings: BuildingComparisonEntry[];
  comparison_dimensions: string[];
  best_passport: string | null;
  worst_passport: string | null;
  average_trust: number;
  average_completeness: number;
}

export const buildingComparisonApi = {
  compare: async (buildingIds: string[]): Promise<BuildingComparison> => {
    const response = await apiClient.post<BuildingComparison>('/buildings/compare', {
      building_ids: buildingIds,
    });
    return response.data;
  },
};

File: backend/app/schemas/building_comparison.py
"""Pydantic v2 schemas for Building Comparison."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class BuildingComparisonRequest(BaseModel):
    """Request body for comparing buildings."""

    building_ids: list[str] = Field(..., min_length=2, max_length=10)

    model_config = ConfigDict(from_attributes=True)


class BuildingComparisonEntry(BaseModel):
    """Comparison data for a single building."""

    building_id: str
    building_name: str
    address: str
    passport_grade: str | None = None
    passport_score: float | None = None
    trust_score: float | None = None
    completeness_score: float | None = None
    readiness_summary: dict[str, bool] = Field(default_factory=dict)
    open_actions_count: int = 0
    open_unknowns_count: int = 0
    contradictions_count: int = 0
    diagnostic_count: int = 0
    last_diagnostic_date: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class BuildingComparison(BaseModel):
    """Aggregated comparison result for multiple buildings."""

    buildings: list[BuildingComparisonEntry]
    comparison_dimensions: list[str]
    best_passport: str | None = None
    worst_passport: str | None = None
    average_trust: float = 0.0
    average_completeness: float = 0.0

    model_config = ConfigDict(from_attributes=True)
